frameworklauncher/src/main/java/com/microsoft/frameworklauncher/webserver/RequestManager.java

```diff
@@ -173,6 +175,53 @@ private void deleteOrphanFrameworks() throws Exception {
     } while (frameworkDeletedInThisPass);
   }
 
+  private void updateExecutionTypeInternal(String frameworkName, ExecutionType executionType) throws Exception {
+    FrameworkRequest frameworkRequest = YamlUtils.deepCopy(
+        checkExist(aggFrameworkRequests.get(frameworkName)).getFrameworkRequest(), FrameworkRequest.class);
+    frameworkRequest.getFrameworkDescriptor().setExecutionType(executionType);
+    setFrameworkRequest(frameworkName, frameworkRequest);
+  }
+
+  private void stopFrameworkRequestInternal(String frameworkName) throws Exception {
+    updateExecutionTypeInternal(frameworkName, ExecutionType.STOP);
+  }
+
+  // stopOrphanFrameworks need to be handled in WebServer side instead of AM side,
+  // since AM is not always running, such as when the FrameworkState is not APPLICATION_RUNNING.
+  private void stopOrphanFrameworks() throws Exception {
+    // A Framework is StopOrphan, if and only if its ParentFramework is not null and Stopped.
```

Comment:
Why “stopOrphanFrameworks" do not exists before?

Reply:
Previously, we only have delete interface, the counterpart is deleteOrphanFrameworks

---
In reply to: [172149640](https://github.com/Microsoft/pai/pull/346#discussion_r172149640) [](ancestors = 172149640)